File: 3-Structural Patterns/4-Composite Pattern/Topic-Lecture-Video Example/Python/composite_pattern_test.py
#!usr/bin/env python3
# -*- coding: utf-8 -*-

"""
Application that actually uses Composite Pattern.
"""

__author__ = 'Ziang Lu'

from model import Lecture, Topic, Video


def main():
    design_patterns = Topic('Design Patterns')

    patterns_intro = Lecture('Intro to Design Patterns')
    design_patterns.add_module(patterns_intro)

    composite = Topic('Composite Pattern')
    design_patterns.add_module(composite)

    composite_intro = Lecture('Intro to Composite Pattern')
    composite.add_module(composite_intro)
    composite_video = Video("Let's compose!")
    composite.add_module(composite_video)

    observer = Topic('Iterator Pattern')
    design_patterns.add_module(observer)

    observer_intro = Lecture('Intro to Observer Pattern')
    observer.add_module(observer_intro)

    design_patterns.display('\t')


if __name__ == '__main__':
    main()

# Output:
# 	  Topic - Design Patterns
#         Lecture - Intro to Design Patterns
#         Topic - Composite Pattern
#                 Lecture - Intro to Composite Pattern
#                 Video - Let's compose!
#         Topic - Observer Pattern
#                 Lecture - Intro to Observer Pattern

File: 3-Structural Patterns/3-Composite Pattern/Topic-Lecture-Video Example/Java/runner/CompositePatternTest.java
package runner;

import model.Lecture;
import model.Topic;
import model.Video;

/**
 * Application that actually uses Composite Pattern.
 *
 * @author Ziang Lu
 */
public class CompositePatternTest {

    /**
     * Main driver.
     * @param args arguments from command line
     */
    public static void main(String[] args) {
        Topic designPatterns = new Topic("Design Patterns");

        Lecture patternsIntro = new Lecture("Intro to Design Patterns");
        designPatterns.addModule(patternsIntro);

        Topic composite = new Topic("Composite Pattern");
        designPatterns.addModule(composite);

        Lecture compositeIntro = new Lecture("Intro to Composite Pattern");
        composite.addModule(compositeIntro);
        Video compositeVideo = new Video("Let's compose!");
        composite.addModule(compositeVideo);

        Topic observer = new Topic("Observer Pattern");
        designPatterns.addModule(observer);

        Lecture observerIntro = new Lecture("Intro to Observer Pattern");
        observer.addModule(observerIntro);

        designPatterns.display("\t");

        /*
         * Output:
         * 	   Topic - Design Patterns
         * 	       Lecture - Intro to Design Patterns
         * 	       Topic - Composite Pattern
         * 	               Lecture - Intro to Composite Pattern
         * 	               Video - Let's compose!
         * 	       Topic - Observer Pattern
         * 	               Lecture - Intro to Observer Pattern
         */
    }

}
